---
title: "Gouldian Finch Genome Seminar"
author: "Hannah Reeb"
date: "2024-12-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
### Project Background 
Gouldian finches are brightly colored birds, with some morphs having yellow plumage on both their belly and head feathers. In those morphs, the yellow color does not come from the same carotenoid pigments in both patches. In the belly feathers, lutein is the source of the yellow, and in the mask, 3'-dehydrolutein is the source. In House Finches, DHRS13 is involved in converting lutein to dehydrolutein, and we have som initial evidence that it is differentially expressed in head versus belly feathers of Gouldian finches. Here, our goal is to assess whether DHRS13 is upregulated in feathers where dehydrolutein is present, implicating its importance in carotenoid conversion in Gouldian finches. We are also interested in finding other differentially expressed genes between the two feather types, and whether any are genes known to be associated with carotenoid metabolism. 

We predict that DHRS13 will be upregulated in head feathers. To quantify this, we take RNA sequences from Gouldian finch head and belly feathers, and map them to a reference genome to find upregulated genes related to carotenoid expression. 

### Data Sources
This analysis uses 8 samples of Gouldian Finch (GOFI) feathers. 4 from the belly, 4 from the head.We use single end, Illumina reads collected in 2018. 

### Methods Overview
1. Fast QC
2. Trim reads
3. Align and count with kallisto
4. Utilize sleuth in R to find DEGs
5. Visualize DEGs using volcano plots and heatmaps
6. Align reads to genome using Hisat2
7. Get BAM file for manipulation
8. Load BAM file to IGV to visualize splicing

## Analysis and Results

## Step 1: FastQC
We will start with an initial survey of read quality using fastqc. I used an array job for this step. Overall, the reads look pretty good-- adapter content was low before trimming, and confidence per base was high. I included one link to the html output from fastQC as an example. 

fastQCsbatch
fastQBargs
fastQCsh
one fast QC html link


## Step 2: Trimming 
We use trimgalore to trim reads. I initially tried an array job for this step, but there were some issues with opening files, which I only discovered after running each job individually. Since that worked best, those are the files I link here. I trim reads to be 100 bp or longer. 

trimming gould sh and sbatch (ALL)

After trimming, adapter content is decreased. We can now have more confidence that our alignments will not be contaminated with adapter sequences!

INSERT PIC

## Step 3: Pseudoalignment with Kallisto
Now that I have trimmed reads, we want to align them to known reference to identify genes of interest. The first step is to create a k-mer indexed reference. I used the gouldian finch cDNA from emsembl.

sbatch sh for index

The next step is to actually align our reads to the indexed reference. This is pseudoalignment. I do this task as an array job. Since we have single end reads, kallisto will not estimate fragment lenght and standard deviation for us. Since we have Illumina reads, I follow the Pachter Lab's recommendation on settings. I use average fragment length = 200, standard deviation of fragment length = 20. Kallisto gives us abundance.tsv and abundance.h5 files as output, which will let us understand up or down regulation of genes later.


pachter lab kallisto 
kallisto sbatch/sh/args files
kallisto output

## Step 4: Sleuthing for DEGs
The sleuth package in R is useful for understanding kallisto output. First, though, we need to pull useful information from the cDNA file we use. I use a regular expression to pull out the Gene ID and Transcrip ID, then manipulate the output in excel until I am left with columsn of Gene ID and Transcript ID to use as a header.

regex for headers
group setup columns
list of headers 

Now that we have organized information and our experimental groups, we can use sleuth to find differentially expressed genes. I used the code Dr. Toomey gave us in class to find them. I try it using data aggregated by gene name, as well as unaggregated. In the unaggregated, every transcript ID is counted by itself. In the aggregated, transcipt IDs with the same gene name are grouped, to give us higher statistical power. 

The volcano plot generated by these analyses uses unaggregated data, as does the heatmap, linked below. 

heatmap
volvanoplot

## Step 5: Uniprot search and diamond blastx 
Oh no! the volcano plot and heat map shows differential gene expression, but the labels are not annotated. That is not of great use to us. 

In order to avoid redoing the last 2 steps of the analysis, I download all of the sequenced proteins for Zebra finches (uniprot), make a database, then query with Gouldian finch DNA using a diamond blastx. 

diamond make db
diamond blast sh/sbatch

From the output, I use a regular expression to pull the headers, and get info that links transcript ID to gene name. 

pic of sample

## Step 6: Redo sleuthing for DEGs
Now, I will repeat Step 4 and retry the sleuth work in R, this time with the zebra finch genome. 


volvano plot 
heatmap

## Step 7: Look at Gene Ontogeny with ShinyGO
I also look at gene ontogeny with ShinyGO

shinygo pic

## Step 8: Search for differentially expressed genes of interest
To start, I make sure that the gene name is what I expect it to be. I do a blastx search with the DHRS13 sequence to Zebra finch, and the name is the same

INSERT SC

From the sleuth analysis, I have a file that lists all significantly upregulated genes in the head feathers, compared to the belly. I can then search by gene names of interest! For this step, transcripts are aggregated by gene name to increase statistical power. 

I enter the following queries:
DHRS
BDH
SCARB
CYP

There is significant upregulation for the following: 
DHRS3: dehydrogenase/reductase 3
SCARB2: scavenger receptor class B member 2
CYP36B1: cytochrome P450 family 26 subfamily B member 1
CYP27A1: cytochrome P450 family 27 subfamily A member 1
CYP27C1: cytochrome P450 family 27 subfamily C member 1
CYP24C1: cytochrome P450 family 24 subfamily A member 1

insert pic wiht upreg nums

## Step 9: Dig deeper on DHRS13
We have foud carotenoid candidate genes that are upregulated in the head, but let's explore further on our gene of highest interest, DHRS13. 

Searching through all genes present, not filtering out those that were significantly differentially expressed, we find the following: 


SC with DHRS13

Differential expression not significant for DHRS13 (p=0.11)
But, it is technically upregulated in the head (Î²=0.45). It has a positive coefficient, meaning that it is more highly expressed in the head, just not to a statistically significant degree. 










## Align trimmed reads to GOFI genome using hisat2

Hisat2 stands for Hierarchical Indexing for Spliced Alignment of Transcripts 2. It allows us to align RNA-Seq reads to a reference genome, which is great for our purposes here (we have transcript reads for Gouldian finch and a reference genome to map to) 

The following code for establishing the hisat2 environment I got from Caleb and Salil. These steps allow us to install hisat2 within an environment using the OSCER login. Adding these channels lets us utilize the necessary packages later in our analysis. 
```{}
$ conda create -n hisat2_env
$ conda activate hisat2_env
$ conda config --add channels bioconda
$ conda config --add channels conda-forge
$ conda config --add channels defaults
$ conda config --show channels
$ conda install hisat2

```

Now, we will use the function "hisat2-build" to create an indexed Gouldian Finch genome that is compatible with Hisat. Hisat is great for this, because unlike some other alignment tools, it takes the fact that introns exist into account during mapping. This works well for samples of RNA-Seq data being aligned to a reference genome, because transcripts will be 'missing' the introns present in the reference genome. Using hisat2 lets us eventually visualize splicing zones and introns.


[index_hisat.sh](/Users/loreeb/Downloads/gouldian finch work/hisat_mapping/index_hisat.sh)
sbatch

Once we have our indexed genome, the next task is to align using hisat2! I did this as an array job. For single end reads, we align each read sample to the indexed genome individually (paired ends get submitted at once as -1 and -2.) Here, we need to specify that we are using single end reads with -U and then the file name of our trimmed reads. 

hisat2 fullmapping sh/sbatch/args 

This step was pretty quick! It took about 9 minutes to run. 
Exciting! Now we have 8 .sam files. 

### .sam to .bam
Our next step is to convert each of these .sam files to .bam files. We will switch back to the mamba class environment for this step, and use samtools view to convert the files. Again, we'll do this step as an array job, since we have many (8) files. 

sam_to_bam sh/sbatch/args

### sort the bam files
Now, we need to sort the bam files by read name. This is what we'll need for the rest of our processing steps

bam_sort sh/sbatch/args

###fixmate namesorted bam files
We will tell samtools to mark the name sorted bam files with tags.

fixmate sh/sbatch/args

### re-sort fixmate
Resort by chromosomal position instead of name. 

resort_fixmate sh/sbathc/args

### Mark and remove duplicates
We use 

markdup sh/sbatch/args

### create index with duplicate removed .bam file
Array job to add index to our bam file. Now, we're ready for stats!

indexed_markdup sh/sbatch/args

### Mapping statistics using flagstats

map_stats sh/sbathc/args
txt files from mapping_stats (scratch)

### Use IGV to visualize splicing


## Conclusions

DHRS13 is not the most significantly differentially expressed carotenoid candidate gene (p=0.11), but it is present in head
Head is where dehydrolutein is expressed
There are other candidate genes with differential expression for head and belly feathers: DHRS3, SCARB2, CYP36B1, CYP27A1, CYP27C1, CYP24C1







